pkg/controller/smart-clone-controller.go

```diff
@@ -1,377 +1,234 @@
-		// more up to date that when the item was initially put onto the
-		// workqueue.
-		if key, ok = obj.(string); !ok {
-			// As the item in the workqueue is actually invalid, we call
-			// Forget here else we'd go into a loop of attempting to
-			// process a work item that is invalid.
-			c.queue.Forget(obj)
-			runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
-			return nil
-		}
-		// Run the syncHandler, passing it the namespace/name string of the data volume
-		// to be synced.
-		if err := c.syncHandler(key); err != nil {
-			// Put the item back on the workqueue to handle any transient errors.
-			c.queue.AddRateLimited(key)
-			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
-		}
-		// Finally, if no error occurs we Forget this item so it does not
-		// get queued again until another change happens.
-		c.queue.Forget(obj)
-		klog.Infof("Successfully synced '%s'", key)
+	// check if volume snapshots exist
+	err := mgr.GetClient().List(context.TODO(), &csiv1.VolumeSnapshotList{})
+	if meta.IsNoMatchError(err) {
 		return nil
-	}(key)
+	}
 
-	if err != nil {
-		runtime.HandleError(err)
-		return true
+	if err != nil && !isErrCacheNotStarted(err) {
+		return err
 	}
 
-	return true
+	if err := importController.Watch(&source.Kind{Type: &csiv1.VolumeSnapshot{}}, &handler.EnqueueRequestForOwner{
+		OwnerType:    &cdiv1.DataVolume{},
+		IsController: true,
+	}, predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return shouldReconcileSnapshot(e.Object.(*csiv1.VolumeSnapshot))
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return shouldReconcileSnapshot(e.ObjectNew.(*csiv1.VolumeSnapshot))
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return shouldReconcileSnapshot(e.Object.(*csiv1.VolumeSnapshot))
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return shouldReconcileSnapshot(e.Object.(*csiv1.VolumeSnapshot))
+		},
+	}); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-func (c *SmartCloneController) syncHandler(key string) error {
-	ns, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		runtime.HandleError(errors.Errorf("invalid resource key: %s", key))
-		return nil
+func shouldReconcileSnapshot(snapshot *csiv1.VolumeSnapshot) bool {
+	_, ok := snapshot.GetAnnotations()[AnnSmartCloneRequest]
+	if !ok {
+		return false
 	}
+	return snapshot.Status.ReadyToUse
+}
 
-	pvc, err := c.pvcLister.PersistentVolumeClaims(ns).Get(name)
-	if err != nil {
-		if k8serrors.IsNotFound(err) {
-			pvc = nil
-		} else {
-			// Error getting PVC - return
-			return err
-		}
-	}
-	snapshot, err := c.snapshotsLister.VolumeSnapshots(ns).Get(name)
-	if err != nil {
-		if k8serrors.IsNotFound(err) {
-			snapshot = nil
-		} else {
-			// Error getting Snapshot - return
-			return err
-		}
+func shouldReconcilePvc(pvc *corev1.PersistentVolumeClaim) bool {
+	if pvc.Status.Phase != corev1.ClaimBound {
```

Comment:
PVCs created by snapshots never waitforfirstconsumer?

Reply:
They can't be, we have written data to it, so if they were WaitForFirstConsumer because they are local storage, then the node has been picked. If it is a topology zone, we have at a minimum written to the storage in that zone, and thus still can't have WaitForFirstConsumer.

Reply:
I was able to simulate this easily in in my local rook ceph configuration by making the storageclass waitforfirstconsumer.

```
➜  kubevirt git:(vmclone) ✗ k get volumesnapshot
NAME            READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS             SNAPSHOTCONTENT                                    CREATIONTIME   AGE
test-snapshot   true         my-pvc                              500Mi         csi-rbdplugin-snapclass   snapcontent-36a79976-79f3-4c5c-825b-90c9b38d5776   7s             9s
```

```
➜  kubevirt git:(vmclone) ✗ k get pvc my-pvc-from-snap -o yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  creationTimestamp: "2020-03-16T14:03:29Z"
  finalizers:
  - kubernetes.io/pvc-protection
  name: my-pvc-from-snap
  namespace: default
  resourceVersion: "2406"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/my-pvc-from-snap
  uid: 7f80ccab-1068-4f67-b715-673f3150f2fd
spec:
  accessModes:
  - ReadWriteOnce
  dataSource:
    apiGroup: snapshot.storage.k8s.io
    kind: VolumeSnapshot
    name: test-snapshot
  resources:
    requests:
      storage: 500Mi
  storageClassName: rook-ceph-block
  volumeMode: Filesystem
status:
  phase: Pending
```

Reply:
Interesting, I didn't think it would be possible. I guess I can add pending to the status that we can reconcile.